dlt/common/libs/pyarrow.py

```diff
@@ -634,124 +649,151 @@ def concat_batches_and_tables_in_order(
-        if col.get("data_type") is None
+    return {
+        column_name: data.ravel() for column_name, data
+        in zip(column_names, np.vsplit(pivoted_rows, len(pivoted_rows)))
     }
 
-    arrow_schema = columns_to_arrow(columns, caps, tz)
 
-    for idx in range(0, len(arrow_schema.names)):
-        field = arrow_schema.field(idx)
-        py_type = type(rows[0][idx])
-        # cast double / float ndarrays to decimals if type mismatch, looks like decimals and floats are often mixed up in dialects
-        if pa.types.is_decimal(field.type) and issubclass(py_type, (str, float)):
-            logger.warning(
-                f"Field {field.name} was reflected as decimal type, but rows contains"
-                f" {py_type.__name__}. Additional cast is required which may slow down arrow table"
-                " generation."
-            )
-            float_array = pa.array(columnar_known_types[field.name], type=pa.float64())
-            columnar_known_types[field.name] = float_array.cast(field.type, safe=False)
-        if issubclass(py_type, (dict, list, set)):
+def convert_numpy_to_arrow(
+    column_data, caps: DestinationCapabilitiesContext, column_schema: TColumnSchema, tz: str
+) -> Any:  # pyarrow.Array
+    """Convert a numpy array to a pyarrow array."""
+    from dlt.common.libs.pyarrow import pyarrow as pa
+
+    dlt_data_type = column_schema.get("data_type")
+    inferred_arrow_type = get_py_arrow_datatype(column_schema, caps, tz) if dlt_data_type is not None else None
+    arrow_array = None
+
+    # base case (0): allow pyarrow to infer type, or create array of specified dlt specified type
+    # NOTE this is more restrictive than `pa.array(column_data).cast(inferred_arrow_type)`
+    try:
+        # type=None lets pyarrow infer the type from the data
+        arrow_array = pa.array(column_data, type=inferred_arrow_type)
+    except (pa.ArrowInvalid, pyarrow.ArrowTypeError) as e:
+        if (
+            "would cause data loss" in e.args[0]  # specific pyarrow error related to precision loss
+            and dlt_data_type == "decimal"
+            and pa.types.is_decimal(inferred_arrow_type)
+        ):
+            raise PyToArrowConversionException(
+                "Received `data_type='decimal'`, but failed to convert to arrow due to insufficient decimal precision and scale. "
+                "Consider setting `precision` and `scale` hints: https://dlthub.com/docs/general-usage/schema/#tables-and-columns"
+            ) from e
+        
+        logger.warning(
+            f"Received `data_type={dlt_data_type}`. Failed to convert to arrow type {inferred_arrow_type}."
+        )
+
+    # case 1: coercing Python types unsupported by Arrow to a JSON friendly format
+    # e.g. dataclasses -> dict, UUID -> str
+    if arrow_array is None and dlt_data_type is None:
+        try:
+            arrow_array = pa.array([
+                None if item is None
+                else custom_encode(item)
```

Comment:
see my comment below. custom encode has misleading signature (it returns json serializable type not always string)

Reply:
feel free to change the signature...